mlflow/sklearn/__init__.py

```diff
@@ -455,6 +458,125 @@ def load_model(model_uri):
+                self.scope_flag_dict[scope_key] = False
+
+        return CallScope()
+
+    def get_run_id_for_model(self, model):
+        return self.model_id_to_run_id_map.get(id(model), None)
+
+    def should_log_eval_metrics(self):
+        return not self.scope_flag_dict[self.fit_scope_key] and \
+            not self.scope_flag_dict[self.eval_and_log_metrics_scope_key] and \
+            not self.scope_flag_dict[self.model_score_scope_key]
+
+    def register_model(self, model, run_id):
+        """
+        In `patched_fit`, we need register the model with the run_id used in `patched_fit`
+        So that in following metric autologging, the metric will be logged into the registered
+        run_id
+        """
+        self.model_id_to_run_id_map[id(model)] = run_id
+
+    def register_prediction_result(self, model, eval_dataset_name, predict_result):
+        """
+        In `patched_predict`, register the prediction result instance with the model id and eval
+        dataset name. e.g.
+
+        ```
+        prediction_result = model_1.predict(eval_X)
+        ```
+        then we need register an item into the `pred_result_id_to_dataset_name_and_model_id` map like
+        below:
+        id(prediction_result) --> (inspected_original_var_name_of(eval_X), id(model_1))
+
+        With this map `dataset_id_to_dataset_name_and_model_id`, in following patched metric API,
+        we can query the eval dataset name and the model id via the `y_pred` argument instance.
+        """
+        value = (eval_dataset_name, id(model))
+        self.pred_result_id_to_dataset_name_and_model_id[id(predict_result)] = value
+
+    def get_eval_dataset_name_and_run_id_from_metric_api_arglist(self, metric_api_call_arg_list):
+        """
+        Given the arguments list of metric API, find the registered dataset name and model id.
+        See `register_prediction_result` method doc for registered information.
+        """
+        # Note: some metric API the arguments is not `y_true`, `y_pred`
+        #  e.g.
+        #    https://scikit-learn.org/stable/modules/generated/sklearn.metrics.roc_auc_score.html#sklearn.metrics.roc_auc_score
+        #    https://scikit-learn.org/stable/modules/generated/sklearn.metrics.silhouette_score.html#sklearn.metrics.silhouette_score
+        dataset_id_list = self.pred_result_id_to_dataset_name_and_model_id.keys()
+        for arg in metric_api_call_arg_list:
+            if id(arg) in dataset_id_list:
+                dataset_name, model_id = self.pred_result_id_to_dataset_name_and_model_id[id(arg)]
+                return dataset_name, self.model_id_to_run_id_map.get(model_id, None)
+        return None, None
+
+
+_autolog_training_status = _AutologTrainingStatus()
+
+
+_metric_api_excluding_list = [
+    'check_scoring', 'get_scorer', 'make_scorer'
+]
+
+
+def _inspect_dataset_var_name(X):
+    name = _inspect_original_var_name(X)
+    if name:
+        return name
+    else:
+        return f'dataset_{id(X)}'
```

Comment:
Does this mean a metric name would look like `accuracy_on_dataset_4475785872`?

Reply:
yes. I am considering shorten the name to be something like "unknown_dataset_{index}"